File: examples/scripts/list_notifications.py
# -*- encoding: utf-8 -*-
"""
SIGNIFY
signify.app.clienting module

Testing clienting with integration tests that require a running KERIA Cloud Agent
"""
import json
from pprint import pprint

from keri.core import coring
from keri.core.coring import Tiers
from keri.core.eventing import messagize, SealEvent
from keri.peer import exchanging
from signify.app.clienting import SignifyClient


def list_notifications():
    url = "http://localhost:3901"
    bran = b'9876543210abcdefghijk'
    tier = Tiers.low

    client = SignifyClient(passcode=bran, tier=tier, url=url)
    identifiers = client.identifiers()
    notificatons = client.notifications()
    groups = client.groups()
    registries = client.registries()

    res = notificatons.list()

    for note in res["notes"]:
        body = note['a']
        route = body['r']
        match route.split("/"):
            case ["", "multisig", "icp"]:
                pass
                # print(body)
                # print(f"Recv: inception request for multisig AID=BLAH")
            case ["", "multisig", "vcp"]:
                said = body['d']
                res = groups.get_request(said=said)
                msg = next(exn for exn in res if exn['exn']['d'] == said)

                sender = msg['sender']
                group = msg["groupName"]

                exn = msg['exn']
                usage = exn['a']["usage"]
                print(f"Credential registry inception request for group AID  {group}:")
                print(f"\tReceived from:  {sender}")
                print(f"\tPurpose:  \"{usage}\"")
                yes = input("Approve [Y|n]? ")

                if yes in ('', 'y', 'Y'):
                    registryName = input("Enter new local name for registry: ")
                    embeds = exn['e']
                    vcp = embeds['vcp']
                    ixn = embeds['ixn']
                    serder = coring.Serder(ked=ixn)
                    ghab = identifiers.get(group)

                    keeper = client.manager.get(aid=ghab)
                    sigs = keeper.sign(ser=serder.raw)

                    ims = messagize(serder=serder, sigers=[coring.Siger(qb64=sig) for sig in sigs])
                    embeds = dict(
                        vcp=coring.Serder(ked=vcp).raw,
                        ixn=ims
                    )

                    sender = ghab["group"]["mhab"]
                    keeper = client.manager.get(aid=sender)
                    exn, end = exchanging.exchange(route="/multisig/vcp",
                                                   payload={'gid': ghab["prefix"], 'usage': "test"},
                                                   sender=sender["prefix"], embeds=embeds)

                    esigs = keeper.sign(ser=exn.raw)
                    groups.send_request(group, exn.ked, esigs, end)

                    return registries.create_from_events(name=group, hab=ghab, registryName=registryName, vcp=vcp,
                                                         ixn=ixn, sigs=sigs)


if __name__ == "__main__":
    list_notifications()

File: examples/scripts/list_notifications.ts
import signify from 'signify-ts';
import promptSync from 'prompt-sync';

const prmpt = promptSync({ sigint: true });

await list_notifications();

async function list_notifications() {
    const url = 'http://127.0.0.1:3901';
    const bran = '0123456789abcdefghijk';

    await signify.ready();
    const client = new signify.SignifyClient(url, bran);
    await client.connect();
    const d = await client.state();
    console.log('Connected: ');
    console.log(' Agent: ', d.agent.i, '   Controller: ', d.controller.state.i);

    const identifiers = client.identifiers();
    const notifications = client.notifications();
    const groups = client.groups();
    const registries = client.registries();

    const res = await notifications.list();
    const notes = res.notes;

    for (const note of notes) {
        const payload = note.a;
        const route = payload.r;

        if (route === '/multisig/vcp') {
            const res = await groups.getRequest(payload.d);
            if (res.length == 0) {
                console.log(
                    'error extracting exns matching nre for ' + payload.data
                );
            }
            const msg = res[0];

            let sender = msg['sender'];
            const group = msg['groupName'];

            const exn = msg['exn'];
            const usage = exn['a']['usage'];
            console.log(
                'Credential registry inception request for group AID  :' + group
            );
            console.log('\tReceived from:  ' + sender);
            console.log('\tPurpose:  ' + usage);
            console.log('\nAuto-creating new registry...');
            const yes = prmpt('Approve [Y|n]? ');
            if (yes === 'y' || yes === 'Y' || yes === '') {
                try {
                    const registryName = prmpt(
                        'Enter new local name for registry: '
                    );
                    let embeds = exn['e'];
                    const vcp = embeds['vcp'];
                    const ixn = embeds['ixn'];
                    const serder = new signify.Serder(ixn);
                    const ghab = await identifiers.get(group);

                    let keeper = client.manager!.get(ghab);
                    const sigs = keeper.sign(signify.b(serder.raw));
                    const sigers = sigs.map(
                        (sig: any) => new signify.Siger({ qb64: sig })
                    );

                    const ims = signify.d(signify.messagize(serder, sigers));
                    const atc = ims.substring(serder.size);
                    embeds = {
                        vcp: [new signify.Serder(vcp), undefined],
                        ixn: [serder, atc],
                    };

                    sender = ghab['group']['mhab'];
                    keeper = client.manager!.get(sender);
                    const [nexn, end] = signify.exchange(
                        '/multisig/vcp',
                        { gid: ghab['prefix'], usage: 'test' },
                        sender['prefix'],
                        undefined,
                        undefined,
                        undefined,
                        undefined,
                        embeds
                    );

                    console.log(nexn.pretty());
                    const esigs = keeper.sign(signify.b(nexn.raw));
                    await groups.sendRequest(
                        group,
                        nexn.ked,
                        esigs,
                        signify.d(end)
                    );

                    return await registries.createFromEvents(
                        ghab,
                        group,
                        registryName,
                        vcp,
                        ixn,
                        sigs
                    );
                } catch (e: any) {
                    console.log(e);
                }
            }
        }
    }
}
